src/H5FDvfd_swmr.c

```diff
@@ -0,0 +1,1945 @@
+    H5FD_vfd_swmr_reader_fapl_t *      new_fa_ptr = NULL;
+    void *                             ret_value  = NULL;
+
+    FUNC_ENTER_PACKAGE
+
+    HDassert(old_fa_ptr);
+    HDassert(old_fa_ptr->magic == H5FD_VFD_SWMR_READER_MAGIC);
+
+    new_fa_ptr = H5FL_CALLOC(H5FD_vfd_swmr_reader_fapl_t);
+    if (NULL == new_fa_ptr)
+        HGOTO_ERROR(H5E_VFL, H5E_CANTALLOC, NULL, "unable to allocate vfd swmr FAPL")
+
+    H5MM_memcpy(new_fa_ptr, old_fa_ptr, sizeof(H5FD_vfd_swmr_reader_fapl_t));
+
+    ret_value = (void *)new_fa_ptr;
+
+done:
+    if (NULL == ret_value) {
+        if (new_fa_ptr) {
+            new_fa_ptr->magic = 0;
+            new_fa_ptr        = H5FL_FREE(H5FD_vfd_swmr_reader_fapl_t, new_fa_ptr);
+        }
+    }
+
+    FUNC_LEAVE_NOAPI(ret_value)
+
+} /* end H5FD__vfd_swmr_fapl_copy() */
+
+/*--------------------------------------------------------------------------
+ * Function:    H5FD__vfd_swmr_fapl_free
+ *
+ * Purpose:     Releases the file access lists
+ *
+ * Return:      SUCCEED/FAIL
+ *--------------------------------------------------------------------------
+ */
+static herr_t
+H5FD__vfd_swmr_fapl_free(void *_fapl)
+{
+    H5FD_vfd_swmr_reader_fapl_t *fapl      = (H5FD_vfd_swmr_reader_fapl_t *)_fapl;
+    herr_t                       ret_value = SUCCEED;
+
+    FUNC_ENTER_PACKAGE
+
+    /* Check arguments */
+    if ((NULL == fapl) || (fapl->magic != H5FD_VFD_SWMR_READER_MAGIC))
+        HGOTO_ERROR(H5E_ARGS, H5E_BADVALUE, FAIL, "NULL or invalid H5FD_vfd_swmr_reader_fapl_t *")
+
+    /* Free the property list */
+    fapl->magic = 0;
+    fapl        = H5FL_FREE(H5FD_vfd_swmr_reader_fapl_t, fapl);
+
+done:
+    FUNC_LEAVE_NOAPI(ret_value)
+
+} /* end H5FD__vfd_swmr_fapl_free() */
+
+/*-------------------------------------------------------------------------
+ * Function:    H5P_pop_vfd_swmr_reader_vfd_off_fapl
+ *
+ * Purpose:     After a file has been opened in VFD SWMR reader mode, we
+ *              must pop the vfd swmr reader driver entry off the supplied
+ *              fapl.  If we don't, and the fapl is used to open a second
+ *              file (i.e. via virtual data sets), we would have multiple
+ *              vfd swmr reader driver entries pushed on the vfd stack.
+ *
+ *              Do this as follows:
+ *
+ *              1) Read the file driver entry from the supplied fapl.  Verify
+ *                 that it specifies the vfd swmr reader VFD.
+ *
+ *              2) Read the file driver entry from the sub fapl specified
+ *                 in the vfd swmr reader vfd fapl entry.  Set the file
+ *                 driver entry on the supplied fapl equal to that on
+ *                 the sub-fapl.
+ *
+ *              3) Discard the sub-fapl?  Not sure if this is necessary.
+ *
+ * Return:      SUCCEED/FAIL
+ *
+ * Programmer   JRM -- 4/28/22
+ *
+ *-------------------------------------------------------------------------
+ */
+herr_t
+H5P_pop_vfd_swmr_reader_vfd_off_fapl(hid_t fapl_id)
```

Comment:
Based on the usage of this routine and H5P_push_vfd_swmr_reader_vfd_on_fapl in the H5F__open code, it appears that the VFD callbacks are being "suspended" and then restored.   This is incorrectly assuming that future developers will understand this hack and is taking on a large technical debt to hack this suspend/restore functionality in, instead of implementing this VFD as a proper pass-through driver.   This code should definitely be removed and the driver refactored to support proper pass-through behavior.